src/orange/univ/detail/LazySenseCalculator.hh

```diff
@@ -0,0 +1,163 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021-2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/univ/detail/LazySenseCalculator.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Assert.hh"
+#include "corecel/cont/Range.hh"
+#include "corecel/cont/Span.hh"
+#include "orange/OrangeTypes.hh"
+#include "orange/surf/LocalSurfaceVisitor.hh"
+#include "orange/univ/detail/Types.hh"
+
+#include "SurfaceFunctors.hh"
+#include "../VolumeView.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate senses with a fixed particle position.
+ *
+ * This is an implementation detail used in initialization *and* complex
+ * intersection.
+ */
+class LazySenseCalculator
+{
+  public:
+    // Construct from persistent, current, and temporary data
+    inline CELER_FUNCTION LazySenseCalculator(LocalSurfaceVisitor const& visit,
+                                              Real3 const& pos,
+                                              Span<Sense> sense_cache,
+                                              Span<SenseModFlags> sense_flags);
+
+    // Calculate senses for a single face of the given volume, possibly on a
+    // face
+    inline CELER_FUNCTION Sense operator()(VolumeView const& vol,
+                                           FaceId face_id,
+                                           OnFace face = {});
+
+    //! Clear the cached sense values
+    CELER_FUNCTION void invalidate_cache()
+    {
+        for (auto& flags : sense_flags_)
+        {
+            flags = unset_sense_mod(SenseMod::cached, flags);
+        }
+    }
+
+    //! The first face encountered that we are "on"
+    CELER_FUNCTION OnFace& on_face() { return face_; }
```

Comment:
Maybe instead the class should take `OnFace&` as a construction argument? The initial value should be left unchanged.

Reply:
>  The initial value should be left unchanged.

Mmh, what do you mean by this?

Reply:
Instead of the function-like class storing and providing an accessor for the face, have it take a pointer to one from the parent scope. It doesn't even have to initialize it on construction: the original "sense calculator" respects the original value (so that it will only override a "false" value if a new face is found).